pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -487,24 +519,78 @@ func (configgen *ConfigGeneratorImpl) buildSidecarOutboundListeners(env *model.E
 				continue
 			}
 
-			if currentListener != nil {
+			// We checked TCP over HTTP, and HTTP over TCP conflicts above.
+			// The code below checks for TCP over TCP conflicts and merges listeners
+			if currentListenerEntry != nil {
 				// merge the newly built listener with the existing listener
-				newFilterChains := make([]listener.FilterChain, 0, len(currentListener.FilterChains)+len(mutable.Listener.FilterChains))
-				newFilterChains = append(newFilterChains, currentListener.FilterChains...)
-				newFilterChains = append(newFilterChains, mutable.Listener.FilterChains...)
-				currentListener.FilterChains = newFilterChains
+				// if and only if the filter chains have distinct conditions
+				// Extract the current filter chain matches
+				// For every new filter chain match being added, check if any previous match is same
+				// if so, skip adding this filter chain with a warning
+				// This is very unoptimized.
+				newFilterChains := make([]listener.FilterChain, 0,
+					len(currentListenerEntry.listener.FilterChains)+len(mutable.Listener.FilterChains))
+				newFilterChains = append(newFilterChains, currentListenerEntry.listener.FilterChains...)
+				for _, incomingFilterChain := range mutable.Listener.FilterChains {
+					conflictFound := false
+
+				compareWithExisting:
+					for _, existingFilterChain := range currentListenerEntry.listener.FilterChains {
+						if existingFilterChain.FilterChainMatch == nil {
+							// This is a catch all filter chain.
+							// We can only merge with a non-catch all filter chain
+							// Else mark it as conflict
+							if incomingFilterChain.FilterChainMatch == nil {
+								conflictFound = true
+								env.PushStatus.Add(model.ProxyStatusConflictOutboundListenerTCPOverTCP,
+									listenerMapKey, node,
+									fmt.Sprintf("Listener=%s TCPServices=%d RejectedTCP=%s",
+										listenerMapKey,
+										len(currentListenerEntry.services),
+										service.Hostname))
+								break compareWithExisting
```

Comment:
wouldn't just `break` be equivalent, since the closest loop is labeled with `compareWithExisting`?